### PR TITLE
[InstCombine] Make `(binop ({s|u}itofp),({s|u}itofp))` transform more flexible to mismatched signs

### PR DESCRIPTION
**Pull request by goldsteinn**
- **[InstCombine] Add more tests for transforming `(binop (uitofp)**
- **[InstCombine] Make `(binop ({s|u}itofp),({s|u}itofp))` transform more flexible to mismatched signs**

### CODE REVIEW

**Comment by dtcxzyw** [comment]
Could you please have a look at https://github.com/dtcxzyw/llvm-opt-benchmark/pull/336#discussion_r1517575694?

**Comment by goldsteinn** [comment]
> Could you please have a look at [dtcxzyw/llvm-opt-benchmark#336 (comment)](https://github.com/dtcxzyw/llvm-opt-benchmark/pull/336#discussion_r1517575694)?

So it seems this transform is enabling SLP vectorization in a case where its not profitable:
```
;; Before
; *** IR Dump After SimplifyCFGPass on regress ***
; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none)
define float @regress(i16 %inp) local_unnamed_addr #0 {
  %r0 = urem i16 %inp, 20
  %ui0 = uitofp i16 %r0 to float
  %fadd0 = fadd float %ui0, -1.000000e+01
  %fdiv0 = fdiv float %fadd0, 0.000000e+00
  %ui1 = uitofp i16 %inp to float
  %fdiv1 = fdiv float %ui1, 0.000000e+00
  %r = fmul float %fdiv1, %fdiv0
  ret float %r
}
; *** IR Dump After SLPVectorizerPass on regress ***
; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none)
define float @regress(i16 %inp) local_unnamed_addr #0 {
  %r0 = urem i16 %inp, 20
  %ui0 = uitofp i16 %r0 to float
  %fadd0 = fadd float %ui0, -1.000000e+01
  %fdiv0 = fdiv float %fadd0, 0.000000e+00
  %ui1 = uitofp i16 %inp to float
  %fdiv1 = fdiv float %ui1, 0.000000e+00
  %r = fmul float %fdiv1, %fdiv0
  ret float %r
}
;; After
; *** IR Dump After SimplifyCFGPass on regress ***
; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none)
define float @regress(i16 %inp) local_unnamed_addr #0 {
  %r0 = urem i16 %inp, 20
  %1 = add nsw i16 %r0, -10
  %fadd0 = sitofp i16 %1 to float
  %fdiv0 = fdiv float %fadd0, 0.000000e+00
  %ui1 = uitofp i16 %inp to float
  %fdiv1 = fdiv float %ui1, 0.000000e+00
  %r = fmul float %fdiv1, %fdiv0
  ret float %r
}
; *** IR Dump After SLPVectorizerPass on regress ***
; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none)
define float @regress(i16 %inp) local_unnamed_addr #0 {
  %r0 = urem i16 %inp, 20
  %1 = add nsw i16 %r0, -10
  %2 = insertelement <2 x i16> poison, i16 %inp, i32 0
  %3 = insertelement <2 x i16> %2, i16 %1, i32 1
  %4 = uitofp <2 x i16> %3 to <2 x float>
  %5 = sitofp <2 x i16> %3 to <2 x float>
  %6 = shufflevector <2 x float> %4, <2 x float> %5, <2 x i32> <i32 0, i32 3>
  %7 = fdiv <2 x float> %6, zeroinitializer
  %8 = extractelement <2 x float> %7, i32 0
  %9 = extractelement <2 x float> %7, i32 1
  %r = fmul float %8, %9
  ret float %r
}
```
looking into a fix.

**Comment by goldsteinn** [comment]
> > > Could you please have a look at [dtcxzyw/llvm-opt-benchmark#336 (comment)](https://github.com/dtcxzyw/llvm-opt-benchmark/pull/336#discussion_r1517575694)?
> > 
> > 
> > So it seems this transform is enabling SLP vectorization in a case where its not profitable:
> > ```
> > ;; Before
> > ; *** IR Dump After SimplifyCFGPass on regress ***
> > ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none)
> > define float @regress(i16 %inp) local_unnamed_addr #0 {
> >   %r0 = urem i16 %inp, 20
> >   %ui0 = uitofp i16 %r0 to float
> >   %fadd0 = fadd float %ui0, -1.000000e+01
> >   %fdiv0 = fdiv float %fadd0, 0.000000e+00
> >   %ui1 = uitofp i16 %inp to float
> >   %fdiv1 = fdiv float %ui1, 0.000000e+00
> >   %r = fmul float %fdiv1, %fdiv0
> >   ret float %r
> > }
> > ; *** IR Dump After SLPVectorizerPass on regress ***
> > ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none)
> > define float @regress(i16 %inp) local_unnamed_addr #0 {
> >   %r0 = urem i16 %inp, 20
> >   %ui0 = uitofp i16 %r0 to float
> >   %fadd0 = fadd float %ui0, -1.000000e+01
> >   %fdiv0 = fdiv float %fadd0, 0.000000e+00
> >   %ui1 = uitofp i16 %inp to float
> >   %fdiv1 = fdiv float %ui1, 0.000000e+00
> >   %r = fmul float %fdiv1, %fdiv0
> >   ret float %r
> > }
> > ;; After
> > ; *** IR Dump After SimplifyCFGPass on regress ***
> > ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none)
> > define float @regress(i16 %inp) local_unnamed_addr #0 {
> >   %r0 = urem i16 %inp, 20
> >   %1 = add nsw i16 %r0, -10
> >   %fadd0 = sitofp i16 %1 to float
> >   %fdiv0 = fdiv float %fadd0, 0.000000e+00
> >   %ui1 = uitofp i16 %inp to float
> >   %fdiv1 = fdiv float %ui1, 0.000000e+00
> >   %r = fmul float %fdiv1, %fdiv0
> >   ret float %r
> > }
> > ; *** IR Dump After SLPVectorizerPass on regress ***
> > ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none)
> > define float @regress(i16 %inp) local_unnamed_addr #0 {
> >   %r0 = urem i16 %inp, 20
> >   %1 = add nsw i16 %r0, -10
> >   %2 = insertelement <2 x i16> poison, i16 %inp, i32 0
> >   %3 = insertelement <2 x i16> %2, i16 %1, i32 1
> >   %4 = uitofp <2 x i16> %3 to <2 x float>
> >   %5 = sitofp <2 x i16> %3 to <2 x float>
> >   %6 = shufflevector <2 x float> %4, <2 x float> %5, <2 x i32> <i32 0, i32 3>
> >   %7 = fdiv <2 x float> %6, zeroinitializer
> >   %8 = extractelement <2 x float> %7, i32 0
> >   %9 = extractelement <2 x float> %7, i32 1
> >   %r = fmul float %8, %9
> >   ret float %r
> > }
> > ```
It seems without the transform, the `fadd` prevents SLP i.e

```
BEFORE:
Operand 0:
    %ui0 = uitofp i16 %r0 to float
    %ui1 = uitofp i16 %inp to float
Operand 1:
  float -1.000000e+01
  float 0.000000e+00
Scalars: 
    %fadd0 = fadd float %ui0, -1.000000e+01
    %fdiv1 = fdiv float %ui1, 0.000000e+00
State: Vectorize
MainOp:   %fadd0 = fadd float %ui0, -1.000000e+01
AltOp:   %fdiv1 = fdiv float %ui1, 0.000000e+00
VectorizedValue: NULL
ReuseShuffleIndices: Empty
ReorderIndices: 
UserTreeIndices: 
SLP: Costs:
SLP:     ReuseShuffleCost = 0
SLP:     VectorCost = 6
SLP:     ScalarCost = 5
SLP:     ReuseShuffleCost + VecCost - ScalarCost = 1
SLP: Adding cost 1 for bundle n=2 [  %fadd0 = fadd float %ui0, -1.000000e+01, ..].
SLP: Current total cost = 1
SLP: Calculated costs for Tree:

AFTER:
0.
Operand 0:
    %fadd0 = sitofp i16 %1 to float
    %ui1 = uitofp i16 %inp to float
Operand 1:
  float 0.000000e+00
  float 0.000000e+00
Scalars: 
    %fdiv0 = fdiv float %fadd0, 0.000000e+00
    %fdiv1 = fdiv float %ui1, 0.000000e+00
State: Vectorize
MainOp:   %fdiv0 = fdiv float %fadd0, 0.000000e+00
AltOp:   %fdiv0 = fdiv float %fadd0, 0.000000e+00
VectorizedValue: NULL
ReuseShuffleIndices: Empty
ReorderIndices: 
UserTreeIndices: 
SLP: Costs:
SLP:     ReuseShuffleCost = 0
SLP:     VectorCost = 4
SLP:     ScalarCost = 8
SLP:     ReuseShuffleCost + VecCost - ScalarCost = -4
SLP: Adding cost -4 for bundle n=2 [  %fdiv0 = fdiv float %fadd0, 0.000000e+00, ..].
SLP: Current total cost = -4
SLP: Calculated costs for Tree:
```

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1413]:
```suggestion
    Constant *Op1FpC, SmallVectorImpl<WithCache<const Value *>> &OpsKnown) {
```
Please pass the container by reference since it is always non-null.

**Review by dtcxzyw — APPROVED** [review]
LGTM w/ a nit. Please wait for additional approval from other reviewers.

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1413]: Definitely prefer the reference. Also, why is this passing IntOps by value?

You might also want to replace the `SmallVector<Value *, 2>` with `std::array<Value *, 2>` here.

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1413]: We overrwrite `IntOps` in the helper.

Will change to std::array/by reference.

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1413]: Fixed.

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1428]:
```suggestion
  // NB: This only comes up if OpsFromSigned is true, so there is no need to
```

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1429]:
```suggestion
  // cache it between calls to `foldFBinOpOfIntCastsFromSign`.
```

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1440]:
```suggestion
    // `isKnownNonNegative`, change this to an explicit call.
```

**Review by nikic — APPROVED** [review]

**Pull request closed without merging**
